RandLAPACK/gpu_functions/rl_cuda_kernels.cuh

```diff
@@ -0,0 +1,934 @@
+    naive_rank_est<<<num_blocks_ger, threadsPerBlock, sizeof(int64_t), stream>>>(n, alpha, A, lda, rank_device);
+#endif
+    // It seems that these synchs are required, after all
+    cudaStreamSynchronize(stream);
+    cudaMemcpy(&rank, rank_device, sizeof(int64_t), cudaMemcpyDeviceToHost);
+    cudaStreamSynchronize(stream);
+    cudaFree(rank_device);
+    cudaError_t ierr = cudaGetLastError();
+    if (ierr != cudaSuccess)
+    {
+        RandLAPACK_CUDA_ERROR("Failed to launch naive_rank_est. " << cudaGetErrorString(ierr))
+        abort();
+    }
+}
+
+template <typename T>
+void all_of(
+    cudaStream_t stream, 
+    int64_t n,
+    const T alpha,
+    const T* A,
+    bool& all_equal
+) {
+#ifdef USE_CUDA
+    bool* all_equal_device;
+    cudaMalloc(&all_equal_device, sizeof(bool));
+
+    constexpr int threadsPerBlock{128};
+    int64_t num_blocks_ger{(n + threadsPerBlock - 1) / threadsPerBlock};
+    all_of<<<num_blocks_ger, threadsPerBlock, sizeof(bool), stream>>>(n, alpha, A, all_equal_device);
+    cudaStreamSynchronize(stream);
+    cudaMemcpy(&all_equal, all_equal_device, sizeof(bool), cudaMemcpyDeviceToHost);
+    cudaStreamSynchronize(stream);
+    cudaFree(all_equal_device);
+#endif
+    cudaError_t ierr = cudaGetLastError();
+    if (ierr != cudaSuccess)
+    {
+        RandLAPACK_CUDA_ERROR("Failed to launch all_equal. " << cudaGetErrorString(ierr))
+        abort();
+    }
+}
+
+template <typename T>
+void ger_gpu(
+    cudaStream_t stream, 
+    int64_t m, 
+    int64_t n, 
+    T alpha, 
+    T *x, 
+    int64_t incx, 
+    T *y, 
+    int64_t incy, 
+    T* A, 
+    int64_t lda
+) {
+
+#ifdef USE_CUDA
+    constexpr int threadsPerBlock{128};
+    int64_t num_blocks_ger{(m + threadsPerBlock - 1) / threadsPerBlock};
+    ger_gpu<<<num_blocks_ger, threadsPerBlock, 0, stream>>>(m, n, alpha, x, incx, y, incy, A, lda);
+#endif
+    cudaError_t ierr = cudaGetLastError();
+    if (ierr != cudaSuccess)
+    {
+        RandLAPACK_CUDA_ERROR("Failed to launch ger_gpu. " << cudaGetErrorString(ierr))
+        abort();
+    }
+}
+
+template <typename T>
+void copy_mat_gpu(
+    cudaStream_t stream, 
+    int64_t m,
+    int64_t n,
+    const T* A,
+    int64_t lda,
+    T* A_cpy,
+    int64_t ldat,
+    bool copy_upper_triangle
+) {
+#ifdef USE_CUDA
+    dim3 threadsPerBlock(32, 8);
+    dim3 numBlocks((n + threadsPerBlock.x - 1) / threadsPerBlock.x,
+                   (m + threadsPerBlock.y - 1) / threadsPerBlock.y);
+    copy_mat_gpu<<<numBlocks, threadsPerBlock, 0, stream>>>(m, n, A, lda, A_cpy, ldat, copy_upper_triangle);
+#endif
+    cudaError_t ierr = cudaGetLastError();
+    if (ierr != cudaSuccess)
+    {
+        RandLAPACK_CUDA_ERROR("Failed to launch copy_mat_gpu. " << cudaGetErrorString(ierr))
+        abort();
+    }
+}
+
+template <typename T>
+void col_swap_gpu(
+    cudaStream_t stream, 
+    int64_t m, 
+    int64_t n, 
+    int64_t k,
+    T* A, 
+    int64_t lda,
+    int64_t const* idx
+) {
+#ifdef USE_CUDA
```

Comment:
This usage of an ifdef is more like what I want to see. But I still want the function to raise an error if it's called and USE_CUDA isn't defined.

Reply:
Fixed.